daceml/distributed/schedule.py

```diff
@@ -0,0 +1,339 @@
+            raise ValueError(f"{exact_size} is not divisible by {block_size}")
+        td_to_new = exact_size // block_size - 1
+        td_step_new = td_step
+        new_ranges.append((0, td_to_new, td_step_new))
+    return subsets.Range(new_ranges)
+
+
+LocalSubsets = Dict[str, subsets.Range]
+RankVariables = List[Optional[symbolic.symbol]]
+
+
+def propagate_rank_local_subsets(
+    sdfg: SDFG, state: SDFGState, map_nodes: MapNodes, num_blocks: NumBlocks
+) -> Tuple[RankVariables, Tuple[LocalSubsets, LocalSubsets]]:
+    """
+    Compute the subset rank local subsets we need.
+
+    For each rank-tiled parameter, the returned dictionary contains a mapping
+    from the original parameter name to the variable name of the rank index.
+
+    The two-tuple contains mappings from array names to the subset expressions
+    for the inputs and outputs
+    """
+    me, mx = map_nodes
+    map_node = me.map
+
+    # We need to reindex using "fake" block indices
+    # Create new symbols for the block indices and block sizes
+    used_vars: Set[str] = set(map_node.params)
+    rank_variables: RankVariables = []
+    # build a new range for the fake local map
+    outer_range = []
+
+    for param, size, (start, end,
+                      step), n_blocks in zip(map_node.params,
+                                             map_node.range.size_exact(),
+                                             map_node.range, num_blocks):
+        if n_blocks != 1:
+            rank_variable = utils.find_str_not_in_set(used_vars,
+                                                      f"__block{param}")
+            used_vars.add(rank_variable)
+            outer_range.append((
+                symbolic.pystr_to_symbolic(
+                    f"{rank_variable} * ({size} / {n_blocks})"),
+                symbolic.pystr_to_symbolic(
+                    f"({rank_variable} + 1) * ({size} / {n_blocks}) - 1"),
+                symbolic.pystr_to_symbolic(f"{step}"),
+            ))
+
+            rank_variables.append(symbolic.pystr_to_symbolic(rank_variable))
+        else:
+            rank_variables.append(symbolic.symbol(node.FULLY_REPLICATED_RANK))
+            outer_range.append((start, end, step))
+
+    outer_range = subsets.Range(outer_range)
+
+    # Collect all defined variables
+    scope_node_symbols = set(conn for conn in me.in_connectors
+                             if not conn.startswith('IN_'))
+    defined_vars = [
+        symbolic.pystr_to_symbolic(s)
+        for s in (state.symbols_defined_at(me).keys()
+                  | sdfg.constants.keys()) if s not in scope_node_symbols
+    ]
+    defined_vars.extend(map(symbolic.pystr_to_symbolic, used_vars))
+    defined_vars = set(defined_vars)
+
+    results: Tuple[LocalSubsets, LocalSubsets] = ({}, {})
+    for is_input, result in zip([True, False], results):
+
+        # gather internal memlets by the out array they write to
+        internal_memlets: Dict[
```

Comment:
Maybe look in scope_subgraph

Reply:
What do you mean? In case there is a global write in the subgraph? Is that allowed?